service/src/Helper.cpp

```diff
@@ -268,4 +270,148 @@ namespace geopm
         }
         return stat_struct.st_gid;
     };
+
+    int pidfd_open(pid_t pid, int flags)
+    {
+        return syscall(SYS_pidfd_open, pid, flags);
+    }
+
+    pid_t pidfd_to_pid(int pidfd)
+    {
+        pid_t pid = -1;
+        int my_pidfd =  -1;
+        try {
+            my_pidfd = geopm::pidfd_open(getpid(), 0);
+            if (my_pidfd == -1) {
+                throw Exception("pidfd_open(): ",
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+            int err = setns(pidfd, CLONE_NEWPID);
+            if (err == -1) {
+                throw Exception("setns(): ",
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+            std::string info_stream_path = "/proc/self/dfinfo/" + std::to_string(pidfd);
+            std::ifstream info_stream(info_stream_path);
+            for (std::string line; std::getline(info_stream, line); ) {
+                std::string key = "Pid:";
+                if (string_begins_with(line, key)) {
+                    pid = std::stoi(line.substr(key.size()));
+                    break;
+                }
+            }
+            if (pid == -1) {
+                throw Exception("pidfd_to_sidpid(): Failed to parse dfinfo file",
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+            err = setns(my_pidfd, CLONE_NEWPID);
```

Comment:
This should probably happen before the preceding `if(..){ throw }`. Or put `getmypidfd/setNS ... resetNS/closemypidfd` inside a scoped wrapper.